tests/test_moisture_sensor.py

```diff
@@ -7,21 +7,38 @@
 
 class MoistureSensorTest(unittest.TestCase):
 
-    def test_moisture_50(self):
+    def test_moisture_read_succeeds_when_adc_succeeds(self):
         """Moisture level read should be the average of the two readings."""
         mock_adc = mock.Mock()
         mock_adc.read_adc.side_effect = [300, (1023 - 400)]
         mock_pi_io = mock.Mock()
         mock_clock = mock.Mock()
         channel = 1
-        gpio_pins = (12, 16)
-        sensor = moisture_sensor.MoistureSensor(mock_adc, mock_pi_io, channel,
-                                                gpio_pins, mock_clock)
-        moisture_level = sensor.moisture()
-
-        self.assertIn(mock.call(12), mock_pi_io.turn_pin_on.call_args_list)
-        self.assertIn(mock.call(12), mock_pi_io.turn_pin_off.call_args_list)
-        self.assertIn(mock.call(16), mock_pi_io.turn_pin_on.call_args_list)
-        self.assertIn(mock.call(16), mock_pi_io.turn_pin_off.call_args_list)
-        mock_clock.wait.assert_called_once()
-        self.assertEqual(moisture_level, 350)
+        gpio_pin_1 = 12
+        gpio_pin_2 = 16
+        sensor = moisture_sensor.MoistureSensor(
+            mock_adc, mock_pi_io, channel, gpio_pin_1, gpio_pin_2, mock_clock)
+
+        self.assertEqual(350, sensor.moisture())
+        self.assertEqual([mock.call(12), mock.call(16)],
+                         mock_pi_io.turn_pin_on.call_args_list)
+        self.assertEqual(
+            [mock.call(16), mock.call(12), mock.call(12), mock.call(16)],
+            mock_pi_io.turn_pin_off.call_args_list)
+        mock_clock.wait.assert_called_once_with(0.1)
+
+    def test_moisture_sensor_pins_turned_off_if_adc_raises_exception(self):
+        mock_adc = mock.Mock()
+        mock_adc.read_adc.side_effect = ValueError()
+        mock_pi_io = mock.Mock()
+        mock_clock = mock.Mock()
+        channel = 1
+        gpio_pin_1 = 12
+        gpio_pin_2 = 16
+        sensor = moisture_sensor.MoistureSensor(
+            mock_adc, mock_pi_io, channel, gpio_pin_1, gpio_pin_2, mock_clock)
+
+        with self.assertRaises(ValueError):
+            sensor.moisture()
+            self.assertEqual([mock.call(16), mock.call(12), mock.call(16)],
+                             mock_pi_io.turn_pin_off.call_args_list)
```

Comment:
Yeah I agree that we should be testing that the pins are left in the "off" state. I like your solution! Minus the large amount of code. Can you reiterate for me why you don't like using ```setUp```? I think the idea is that someone might only be interested in reading the code for some tests, and not others, and it makes it more difficult to have to scroll to ```setUp``` to figure out what's going on? In a case like this, it seems to me like it's a good way to share code because otherwise we're kind of clogging the tests anyway, and if we only have two tests it isn't difficult to get to the top of the file and look at ```setUp```. 

In the meantime I'll try to think of other ways to do this more concisely.